File: Design/alog/first/数据结构/二叉树/py/二叉树的层序遍历.py
# Definition for a binary tree node.
import queue
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = []

        que = queue.Queue()
        que.put(root)
        while not que.empty():
            level = []
            for i in range(que.qsize()):
                cur = que.get()
                level.append(cur.val)

                if cur.left:
                    que.put(cur.left)
                if cur.right:
                    que.put(cur.right)

            ans.append(level)

        return ans

File: Design/alog/hot100/first/数据结构/二叉树/ts/二叉树的层序遍历.ts
import { TreeNode } from "./types";

function levelOrder(root: TreeNode | null): number[][] {
  if (root === null) {
    return [];
  }
  const res: number[][] = [];
  const q: TreeNode[] = [root];

  while (q.length) {
    const level: number[] = []
    const len = q.length; // 避免length的动态变化
    for (let i = 0; i < len; i++) {
      const top = q.shift()!;
      level.push(top.val);

      if (top.left) q.push(top.left);
      if (top.right) q.push(top.right);
    }
    res.push([...level])
  }

  return res;
};